File: src/main/scala/com/dyonovan/neotech/common/tiles/machines/processors/TileAlloyer.scala
package com.dyonovan.neotech.common.tiles.machines.processors

import com.dyonovan.neotech.client.gui.machines.processors.GuiAlloyer
import com.dyonovan.neotech.common.container.machines.processors.ContainerAlloyer
import com.dyonovan.neotech.common.tiles.MachineProcessor
import com.dyonovan.neotech.managers.{MetalManager, RecipeManager}
import com.dyonovan.neotech.registries.AlloyerRecipeHandler
import com.teambr.bookshelf.client.gui.{GuiColor, GuiTextFormat}
import com.teambr.bookshelf.common.tiles.traits.FluidHandler
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.item.ItemStack
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.util.{EnumFacing, EnumParticleTypes, StatCollector}
import net.minecraft.world.World
import net.minecraftforge.fluids.{Fluid, FluidStack, FluidTank, IFluidHandler}

/**
  * This file was created for NeoTech
  *
  * NeoTech is licensed under the
  * Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License:
  * http://creativecommons.org/licenses/by-nc-sa/4.0/
  *
  * @author Paul Davis <pauljoda>
  * @since 2/20/2016
  */
class TileAlloyer extends MachineProcessor[(FluidStack, FluidStack), FluidStack] with FluidHandler {

    lazy val INPUT_TANK_1 = 0
    lazy val INPUT_TANK_2 = 1
    lazy val OUTPUT_TANK  = 2

    lazy val BASE_ENERGY_TICK = 100

    /**
      * Used to get how long it takes to cook things, you should check for upgrades at this point
      *
      * @return The time it takes in ticks to cook the current item
      */
    override def getCookTime : Int = {
        if(getUpgradeBoard != null && getUpgradeBoard.getProcessorCount > 0)
            1000 - (getUpgradeBoard.getProcessorCount * 112)
        else
            1000
    }

    /**
      * Used to tell if this tile is able to process
      *
      * @return True if you are able to process
      */
    override def canProcess: Boolean = {
        if (energyStorage.getEnergyStored >= getEnergyCostPerTick) {
            return tanks(INPUT_TANK_1).getFluid != null && tanks(INPUT_TANK_2) != null &&
                    RecipeManager.getHandler[AlloyerRecipeHandler](RecipeManager.Alloyer)
                            .isValidInput(tanks(INPUT_TANK_1).getFluid, tanks(INPUT_TANK_2).getFluid) &&
                    (if(tanks(OUTPUT_TANK).getFluid == null) true else RecipeManager.getHandler[AlloyerRecipeHandler](RecipeManager.Alloyer)
                            .getOutput(tanks(INPUT_TANK_1).getFluid, tanks(INPUT_TANK_2).getFluid).get.amount + tanks(OUTPUT_TANK).getFluidAmount <= tanks(OUTPUT_TANK).getCapacity)
        }
        false
    }
    /**
      * Get the output of the recipe
      *
      * @param input The input
      * @return The output
      */
    override def getOutput(input: (FluidStack, FluidStack)): FluidStack =
        if(RecipeManager.getHandler[AlloyerRecipeHandler](RecipeManager.Alloyer).getOutput(input).isDefined)
            RecipeManager.getHandler[AlloyerRecipeHandler](RecipeManager.Alloyer).getOutput(input).get
        else
            null

    /**
      * Get the output of the recipe (used in insert options)
      *
      * @param input The input
      * @return The output
      */
    override def getOutputForStack(input: ItemStack): ItemStack = null

    /**
      * Used to actually cook the item
      */
    override def cook(): Unit = cookTime += 1

    /**
      * Called when the tile has completed the cook process
      */
    override def completeCook(): Unit = {
        if(canProcess) { //Just to be safe
        val recipeTest = RecipeManager.getHandler[AlloyerRecipeHandler](RecipeManager.Alloyer).getRecipe((tanks(INPUT_TANK_1).getFluid, tanks(INPUT_TANK_2).getFluid))
            if(recipeTest.isDefined) {
                val recipe = recipeTest.get
                val output = recipe.getOutput((tanks(INPUT_TANK_1).getFluid, tanks(INPUT_TANK_2).getFluid))
                //Drain Inputs
                val drain1 = tanks(INPUT_TANK_1).drain(recipe.getFluidFromString(recipe.fluidOne).amount, false)
                val drain2 = tanks(INPUT_TANK_2).drain(recipe.getFluidFromString(recipe.fluidTwo).amount, false)

                if(drain1 != null && drain2 != null && drain1.amount > 0 && drain2.amount > 0) {
                    tanks(INPUT_TANK_1).drain(recipe.getFluidFromString(recipe.fluidOne).amount, true)
                    tanks(INPUT_TANK_2).drain(recipe.getFluidFromString(recipe.fluidTwo).amount, true)
                    tanks(OUTPUT_TANK).fill(output.get, true)
                }
            }
        }
    }

    /**
      * Used to get how much energy to drain per tick, you should check for upgrades at this point
      *
      * @return How much energy to drain per tick
      */
    override def getEnergyCostPerTick: Int =
        if(getUpgradeBoard != null && getUpgradeBoard.getProcessorCount > 0)
            BASE_ENERGY_TICK * getUpgradeBoard.getProcessorCount
        else
            BASE_ENERGY_TICK

    /**
      * The initial size of the inventory
      *
      * @return
      */
    override def initialSize: Int = 0

    /**
      * Used to get what slots are allowed to be input
      *
      * @return The slots to input from
      */
    override def getInputSlots: Array[Int] = Array()

    /**
      * Used to get what slots are allowed to be output
      *
      * @return The slots to output from
      */
    override def getOutputSlots: Array[Int] = Array()

    override def writeToNBT(tag : NBTTagCompound) : Unit = {
        super[MachineProcessor].writeToNBT(tag)
        super[FluidHandler].writeToNBT(tag)
    }

    override def readFromNBT(tag : NBTTagCompound) : Unit = {
        super[MachineProcessor].readFromNBT(tag)
        super[FluidHandler].readFromNBT(tag)
    }

    /*******************************************************************************************************************
      **************************************************** Fluid methods ***********************************************
      ******************************************************************************************************************/

    override def setupTanks(): Unit = {
        tanks += new FluidTank(10 * MetalManager.BLOCK_MB) // IN 1
        tanks += new FluidTank(10 * MetalManager.BLOCK_MB) // IN 2
        tanks += new FluidTank(10 * MetalManager.BLOCK_MB) // OUT
    }

    override def onTankChanged(tank: FluidTank): Unit = worldObj.markBlockForUpdate(pos)

    override def getOutputTanks: Array[Int] = Array(OUTPUT_TANK)

    override def getInputTanks: Array[Int] = Array(INPUT_TANK_1, INPUT_TANK_2)

    /**
      * Returns true if the given fluid can be inserted into the given direction.
      *
      * More formally, this should return true if fluid is able to enter from the given direction.
      */
    override def canFill(from: EnumFacing, fluid: Fluid): Boolean = {
        if(fluid == null) return false
        if(tanks(INPUT_TANK_1).getFluid == null)
            return RecipeManager.getHandler[AlloyerRecipeHandler](RecipeManager.Alloyer).isValidSingle(new FluidStack(fluid, 1000))
        else if(tanks(INPUT_TANK_2).getFluid == null)
            return RecipeManager.getHandler[AlloyerRecipeHandler](RecipeManager.Alloyer).isValidSingle(new FluidStack(fluid, 1000))
        else {
            if(fluid == tanks(INPUT_TANK_1).getFluid.getFluid || fluid == tanks(INPUT_TANK_2).getFluid.getFluid)
                return true
            else
                return false
        }
        false
    }

    /**
      * This will try to take things from other inventories and put it into ours
      */
    override def tryInput() : Unit = {
        super.tryInput()
        for(dir <- EnumFacing.values) {
            if(canInputFromSide(dir)) {
                worldObj.getTileEntity(pos.offset(dir)) match {
                    case otherTank : IFluidHandler =>
                        if(otherTank.getTankInfo(dir.getOpposite) != null && otherTank.getTankInfo(dir.getOpposite).nonEmpty &&
                                otherTank.getTankInfo(dir.getOpposite)(0) != null && otherTank.getTankInfo(dir.getOpposite)(0).fluid != null && canFill(dir, otherTank.getTankInfo(dir.getOpposite)(0).fluid.getFluid)) {
                            val amount = fill(dir, otherTank.drain(dir.getOpposite, 1000, false), doFill = false)
                            if (amount > 0)
                                fill(dir, otherTank.drain(dir.getOpposite, amount, true), doFill = true)
                        }
                    case _ =>
                }
            }
        }
    }

    /**
      * This will try to take things from other inventories and put it into ours
      */
    override def tryOutput() : Unit = {
        super.tryOutput()
        for(dir <- EnumFacing.values) {
            if(canOutputFromSide(dir)) {
                worldObj.getTileEntity(pos.offset(dir)) match {
                    case otherTank : IFluidHandler =>
                        val fluid = otherTank.getTankInfo(dir.getOpposite)(0).fluid
                        if((if(fluid == null) true else
                        if (tanks(OUTPUT_TANK).getFluid != null)
                            otherTank.getTankInfo(dir.getOpposite)(0).fluid.getFluid == tanks(OUTPUT_TANK).getFluid.getFluid
                        else false)
                                && canDrain(dir.getOpposite, if(fluid != null) fluid.getFluid else null)) {
                            val amount = otherTank.fill(dir.getOpposite, drain(dir, 1000, doDrain = false), false)
                            if (amount > 0)
                                otherTank.fill(dir, drain(dir.getOpposite, amount, doDrain = true),  true)
                        }
                    case _ =>
                }
            }
        }
    }

    /*******************************************************************************************************************
      ***************************************************** Misc methods ***********************************************
      ******************************************************************************************************************/

    override def getDescription : String = {
        GuiColor.YELLOW + "" + GuiTextFormat.BOLD + StatCollector.translateToLocal("tile.neotech:alloyer.name") + ":\n" +
                GuiColor.WHITE + StatCollector.translateToLocal("neotech.alloyer.desc") + "\n\n" +
                GuiColor.GREEN + GuiTextFormat.BOLD + GuiTextFormat.UNDERLINE + StatCollector.translateToLocal("neotech.text.upgrades") + ":\n" + GuiTextFormat.RESET +
                GuiColor.YELLOW + GuiTextFormat.BOLD + StatCollector.translateToLocal("neotech.text.processors") + ":\n" +
                GuiColor.WHITE + StatCollector.translateToLocal("neotech.electricCrucible.processorUpgrade.desc") + "\n\n" +
                GuiColor.YELLOW + GuiTextFormat.BOLD + StatCollector.translateToLocal("neotech.text.hardDrives") + ":\n" +
                GuiColor.WHITE + StatCollector.translateToLocal("neotech.electricFurnace.hardDriveUpgrade.desc") + "\n\n" +
                GuiColor.YELLOW + GuiTextFormat.BOLD + StatCollector.translateToLocal("neotech.text.control") + ":\n" +
                GuiColor.WHITE + StatCollector.translateToLocal("neotech.electricFurnace.controlUpgrade.desc") + "\n\n" +
                GuiColor.YELLOW + GuiTextFormat.BOLD + StatCollector.translateToLocal("neotech.text.expansion") + ":\n" +
                GuiColor.WHITE +  StatCollector.translateToLocal("neotech.electricFurnace.expansionUpgrade.desc")
    }

    /**
      * Return the container for this tile
      *
      * @param ID Id, probably not needed but could be used for multiple guis
      * @param player The player that is opening the gui
      * @param world The world
      * @param x X Pos
      * @param y Y Pos
      * @param z Z Pos
      * @return The container to open
      */
    override def getServerGuiElement(ID: Int, player: EntityPlayer, world: World, x: Int, y: Int, z: Int): AnyRef =
        new ContainerAlloyer(player.inventory, this)

    /**
      * Return the gui for this tile
      *
      * @param ID Id, probably not needed but could be used for multiple guis
      * @param player The player that is opening the gui
      * @param world The world
      * @param x X Pos
      * @param y Y Pos
      * @param z Z Pos
      * @return The gui to open
      */
    override def getClientGuiElement(ID: Int, player: EntityPlayer, world: World, x: Int, y: Int, z: Int): AnyRef =
        new GuiAlloyer(player, this)

    /**
      * Used to output the redstone single from this structure
      *
      * Use a range from 0 - 16.
      *
      * 0 Usually means that there is nothing in the tile, so take that for lowest level. Like the generator has no energy while
      * 16 is usually the flip side of that. Output 16 when it is totally full and not less
      *
      * @return int range 0 - 16
      */
    override def getRedstoneOutput: Int = if(isActive) 16 else 0

    /**
      * Used to get what particles to spawn. This will be called when the tile is active
      */
    override def spawnActiveParticles(x: Double, y: Double, z: Double): Unit = {
        worldObj.spawnParticle(EnumParticleTypes.SMOKE_NORMAL, x, y, z, 0, 0, 0)
        worldObj.spawnParticle(EnumParticleTypes.SMOKE_NORMAL, x, y, z, 0, 0, 0)
        worldObj.spawnParticle(EnumParticleTypes.SMOKE_NORMAL, x, y, z, 0, 0, 0)
    }
}

File: src/main/java/com/teambrmodding/neotech/common/tiles/machines/processors/TileAlloyer.java
package com.teambrmodding.neotech.common.tiles.machines.processors;

import com.teambr.bookshelf.client.gui.GuiColor;
import com.teambr.bookshelf.client.gui.GuiTextFormat;
import com.teambr.bookshelf.util.ClientUtils;
import com.teambrmodding.neotech.client.gui.machines.processors.GuiAlloyer;
import com.teambrmodding.neotech.collections.EnumInputOutputMode;
import com.teambrmodding.neotech.common.container.machines.processors.ContainerAlloyer;
import com.teambrmodding.neotech.common.tiles.MachineProcessor;
import com.teambrmodding.neotech.common.tiles.traits.IUpgradeItem;
import com.teambrmodding.neotech.managers.MetalManager;
import com.teambrmodding.neotech.managers.RecipeManager;
import com.teambrmodding.neotech.registries.AlloyerRecipeHandler;
import com.teambrmodding.neotech.registries.recipes.AbstractRecipe;
import com.teambrmodding.neotech.registries.recipes.AlloyerRecipe;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.ItemStack;
import net.minecraft.util.EnumFacing;
import net.minecraft.util.EnumParticleTypes;
import net.minecraft.world.World;
import net.minecraftforge.fluids.Fluid;
import net.minecraftforge.fluids.FluidStack;
import net.minecraftforge.fluids.FluidTank;
import net.minecraftforge.fluids.capability.CapabilityFluidHandler;
import net.minecraftforge.fluids.capability.IFluidHandler;
import net.minecraftforge.items.IItemHandler;
import org.apache.commons.lang3.tuple.Pair;

import javax.annotation.Nullable;
import java.util.ArrayList;
import java.util.List;

/**
 * This file was created for NeoTech
 *
 * NeoTech is licensed under the
 * Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License:
 * http://creativecommons.org/licenses/by-nc-sa/4.0/
 *
 * @author Paul Davis - pauljoda
 * @since 2/16/2017
 */
public class TileAlloyer extends MachineProcessor<Pair<FluidStack, FluidStack>, FluidStack> {
    // Class Variables
    public static final int BASE_ENERGY_TICK = 300;
    public static final int INPUT_TANK_1 = 0;
    public static final int INPUT_TANK_2 = 1;
    public static final int OUTPUT_TANK  = 2;

    /**
     * The initial size of the inventory
     */
    @Override
    public int getInventorySize() {
        return 0;
    }

    /**
     * Add all modes you want, in order, here
     */
    @Override
    public void addValidModes() {
        validModes.add(EnumInputOutputMode.INPUT_ALL);
        validModes.add(EnumInputOutputMode.INPUT_PRIMARY);
        validModes.add(EnumInputOutputMode.INPUT_SECONDARY);
        validModes.add(EnumInputOutputMode.OUTPUT_ALL);
        validModes.add(EnumInputOutputMode.ALL_MODES);
    }

    /**
     * Return the list of upgrades by their id that are allowed in this machine
     *
     * @return A list of valid upgrades
     */
    @Override
    public List<String> getAcceptableUpgrades() {
        List<String> list = new ArrayList<>();
        list.add(IUpgradeItem.CPU_SINGLE_CORE);
        list.add(IUpgradeItem.CPU_DUAL_CORE);
        list.add(IUpgradeItem.CPU_QUAD_CORE);
        list.add(IUpgradeItem.CPU_OCT_CORE);
        list.add(IUpgradeItem.MEMORY_DDR1);
        list.add(IUpgradeItem.MEMORY_DDR2);
        list.add(IUpgradeItem.MEMORY_DDR3);
        list.add(IUpgradeItem.MEMORY_DDR4);
        list.add(IUpgradeItem.PSU_250W);
        list.add(IUpgradeItem.PSU_500W);
        list.add(IUpgradeItem.PSU_750W);
        list.add(IUpgradeItem.PSU_960W);
        list.add(IUpgradeItem.TRANSFORMER);
        list.add(IUpgradeItem.REDSTONE_CIRCUIT);
        list.add(IUpgradeItem.NETWORK_CARD);
        return list;
    }

    /**
     * Used to get how much energy to drain per tick, you should check for upgrades at this point
     *
     * @return How much energy to drain per tick
     */
    @Override
    public int getEnergyCostPerTick() {
        return BASE_ENERGY_TICK * getModifierForCategory(IUpgradeItem.ENUM_UPGRADE_CATEGORY.MEMORY) +
                ((getModifierForCategory(IUpgradeItem.ENUM_UPGRADE_CATEGORY.CPU) - 1) * 62);
    }

    /**
     * Used to get how long it takes to cook things, you should check for upgrades at this point
     *
     * @return The time it takes in ticks to cook the current item
     */
    @Override
    public int getCookTime() {
        return Math.max(500 - (((500 / 16) * getModifierForCategory(IUpgradeItem.ENUM_UPGRADE_CATEGORY.CPU) - 1)), 5);
    }

    /**
     * Used to tell if this tile is able to process
     *
     * @return True if you are able to process
     */
    @Override
    public boolean canProcess() {
        if(energyStorage.getEnergyStored() >= getEnergyCostPerTick()) {
            return tanks[INPUT_TANK_1].getFluid() != null && tanks[INPUT_TANK_2].getFluid() != null && // Has inputs
                    RecipeManager.getHandler(RecipeManager.RecipeType.ALLOYER)
                            .isValidInput(Pair.of(tanks[INPUT_TANK_1].getFluid(), tanks[INPUT_TANK_2].getFluid())) && // Is Recipe
                    (tanks[OUTPUT_TANK].getFluid() == null || // Has nothing in output
                    ((AlloyerRecipeHandler) RecipeManager.getHandler(RecipeManager.RecipeType.ALLOYER))
                            .getOutput(Pair.of(tanks[INPUT_TANK_1].getFluid(), tanks[INPUT_TANK_2].getFluid())).amount +
                            tanks[OUTPUT_TANK].getFluidAmount() <= tanks[OUTPUT_TANK].getCapacity()); // Can fill output
        }
        failCoolDown = 40;
        return false;
    }

    /**
     * Used to actually cook the item
     */
    @Override
    protected void cook() {
        cookTime++;
    }

    /**
     * Called when the tile has completed the cook process
     */
    @Override
    protected void completeCook() {
        for(int x = 0; x < getModifierForCategory(IUpgradeItem.ENUM_UPGRADE_CATEGORY.MEMORY); x++) {
            if(canProcess()) {
                AlloyerRecipe recipe =
                        ((AlloyerRecipeHandler)RecipeManager.getHandler(RecipeManager.RecipeType.ALLOYER))
                                .getRecipe(Pair.of(tanks[INPUT_TANK_1].getFluid(), tanks[INPUT_TANK_2].getFluid()));
                if(recipe != null) {
                    // Drain the inputs
                    FluidStack drainInputOne;
                    FluidStack drainOutputTwo;

                    // Find Recipe Fluid Stack One
                    if(tanks[INPUT_TANK_1].getFluid().getFluid() == AbstractRecipe.getFluidStackFromString(recipe.fluidStackOne).getFluid())
                        drainInputOne = tanks[INPUT_TANK_1].drain(AbstractRecipe.getFluidStackFromString(recipe.fluidStackOne).amount, false);
                    else
                        drainInputOne = tanks[INPUT_TANK_1].drain(AbstractRecipe.getFluidStackFromString(recipe.fluidStackTwo).amount, false);

                    // Find Recipe Fluid Stack Two
                    if(tanks[INPUT_TANK_2].getFluid().getFluid() == AbstractRecipe.getFluidStackFromString(recipe.fluidStackTwo).getFluid())
                        drainOutputTwo = tanks[INPUT_TANK_2].drain(AbstractRecipe.getFluidStackFromString(recipe.fluidStackTwo).amount, false);
                    else
                        drainOutputTwo = tanks[INPUT_TANK_2].drain(AbstractRecipe.getFluidStackFromString(recipe.fluidStackOne).amount, false);

                    if(drainInputOne != null && drainOutputTwo != null && drainInputOne.amount > 0 && drainOutputTwo.amount > 0) {
                        tanks[INPUT_TANK_1].drain(drainInputOne, true);
                        tanks[INPUT_TANK_2].drain(drainOutputTwo, true);
                        tanks[OUTPUT_TANK].fill(AbstractRecipe.getFluidStackFromString(recipe.fluidStackOutput), true);
                    }
                }
            }
            else
                break;
        }
        markForUpdate(6);
    }

    /**
     * Get the output of the recipe
     *
     * @param input The input
     * @return The output
     */
    @Override
    public FluidStack getOutput(Pair<FluidStack, FluidStack> input) {
        return ((AlloyerRecipeHandler)RecipeManager.getHandler(RecipeManager.RecipeType.ALLOYER)).getOutput(input);
    }

    /**
     * Get the output of the recipe (used in insert options)
     *
     * @param input The input
     * @return The output
     */
    @Override
    public ItemStack getOutputForStack(ItemStack input) {
        return null;
    }

    /*******************************************************************************************************************
     * Tile Methods                                                                                                    *
     *******************************************************************************************************************/

    /**
     * This will try to take things from other inventories and put it into ours
     */
    @Override
    public void tryInput() {
        for(EnumFacing dir : EnumFacing.values()) {
            if(!isDisabled(dir)) {
                if(world.getTileEntity(pos.offset(dir)) != null &&
                        world.getTileEntity(pos.offset(dir)).hasCapability(CapabilityFluidHandler.FLUID_HANDLER_CAPABILITY, dir.getOpposite())) {
                    IFluidHandler otherTank = world.getTileEntity(pos.offset(dir)).getCapability(CapabilityFluidHandler.FLUID_HANDLER_CAPABILITY, dir.getOpposite());

                    // Attempt fill left tank
                    if(canInputFromSide(dir, true)) {
                        // Try and match our fluid
                        if(tanks[INPUT_TANK_1].getFluid() != null &&
                                otherTank.drain(tanks[INPUT_TANK_1].getFluid(), false) != null) {
                            int amount = tanks[INPUT_TANK_1].fill(otherTank.drain(1000, false), false);
                            if(amount > 0) {
                                tanks[INPUT_TANK_1].fill(otherTank.drain(amount, true), true);
                                markForUpdate(6);
                            }
                        }

                        // Check if we can accept other fluid
                        else if(tanks[INPUT_TANK_1].getFluid() == null && otherTank.drain(1000, false) != null) {
                            FluidStack drainedFluid = otherTank.drain(1000, false);
                            boolean hasExistingOtherFluid = tanks[INPUT_TANK_2].getFluid() != null;
                            int amount = tanks[INPUT_TANK_1].fill(drainedFluid, false);
                            if(amount > 0 &&
                                    (!hasExistingOtherFluid || RecipeManager.getHandler(RecipeManager.RecipeType.ALLOYER)
                                            .isValidInput(Pair.of(drainedFluid, tanks[INPUT_TANK_2].getFluid())))) {
                                tanks[INPUT_TANK_1].fill(otherTank.drain(amount, true), true);
                                markForUpdate(6);
                            }
                        }
                    }

                    // Right Tank
                    if(canInputFromSide(dir, false)) {
                        // Try and match existing
                        if(tanks[INPUT_TANK_2].getFluid() != null && otherTank.drain(tanks[INPUT_TANK_2].getFluid(), false) != null) {
                            int amount = tanks[INPUT_TANK_2].fill(otherTank.drain(1000, false), false);
                            if(amount > 0) {
                                tanks[INPUT_TANK_2].fill(otherTank.drain(amount, true), true);
                                markForUpdate(6);
                            }
                        }

                        // Check if applicable
                        else if(tanks[INPUT_TANK_2].getFluid() == null && otherTank.drain(1000, false) != null) {
                            FluidStack drainedStack = otherTank.drain(1000, false);
                            boolean hasExistingInput = tanks[INPUT_TANK_1].getFluid() != null;
                            int amount = tanks[INPUT_TANK_2].fill(drainedStack, false);
                            if(amount > 0 &&
                                    (!hasExistingInput || RecipeManager.getHandler(RecipeManager.RecipeType.ALLOYER)
                                            .isValidInput(Pair.of(drainedStack, tanks[INPUT_TANK_1].getFluid())))) {
                                tanks[INPUT_TANK_2].fill(otherTank.drain(amount, true), true);
                                markForUpdate(6);
                            }
                        }
                    }
                }
            }
        }
    }

    /**
     * This will try to take things from our inventory and try to place them in others
     */
    @Override
    public void tryOutput() {
        for(EnumFacing dir : EnumFacing.values()) {
            if(canOutputFromSide(dir, true)) {
                if(world.getTileEntity(pos.offset(dir)) != null &&
                        world.getTileEntity(pos.offset(dir)).hasCapability(CapabilityFluidHandler.FLUID_HANDLER_CAPABILITY, dir.getOpposite())) {
                    IFluidHandler otherTank =
                            world.getTileEntity(pos.offset(dir)).getCapability(CapabilityFluidHandler.FLUID_HANDLER_CAPABILITY, dir.getOpposite());

                    // If we have something, try move out
                    if(tanks[OUTPUT_TANK].getFluid() != null && otherTank.fill(tanks[OUTPUT_TANK].getFluid(), false) > 0) {
                        FluidStack amount = drain(otherTank.fill(tanks[OUTPUT_TANK].getFluid(), false), false);
                        if(amount != null) {
                            drain(otherTank.fill(amount, true), true);
                            markForUpdate(6);
                        }
                    }
                }
            }
        }
    }

    /**
     * Used to expose the item handler capability, in child classes, manage input and output based on this
     *
     * @param dir The direction
     * @return The item handler
     */
    @Nullable
    @Override
    protected IItemHandler getItemHandlerCapability(EnumFacing dir) {
        return null;
    }

    /**
     * Used to get the fluid handler for exposing
     *
     * @param dir The direction
     * @return The fluid handler
     */
    @Nullable
    @Override
    protected IFluidHandler getFluidHandlerCapability(EnumFacing dir) {
        return getModeForSide(dir) == EnumInputOutputMode.DEFAULT ||
                canInputFromSide(dir, true) || canInputFromSide(dir, false) || canOutputFromSide(dir, true) ?
                super.getFluidHandlerCapability(dir) : null;
    }

    /*******************************************************************************************************************
     * Inventory Methods                                                                                               *
     *******************************************************************************************************************/

    /**
     * Used to get what slots are allowed to be output
     *
     * @param mode
     * @return The slots to output from
     */
    @Override
    public int[] getOutputSlots(EnumInputOutputMode mode) {
        return new int[0];
    }

    /**
     * Used to get what slots are allowed to be input
     *
     * @param mode
     * @return The slots to input from
     */
    @Override
    public int[] getInputSlots(EnumInputOutputMode mode) {
        return new int[0];
    }

    /*******************************************************************************************************************
     * FluidHandler Methods                                                                                            *
     *******************************************************************************************************************/

    /**
     * Method to define if this tile is a fluid handler
     *
     * @return False by default, override in child classes to enable fluid handling
     */
    @Override
    public boolean isFluidHandler() {
        return true;
    }

    /**
     * Used to set up the tanks needed. You can insert any number of tanks
     *
     * MUST OVERRIDE IN CHILD CLASSES IF isFluidHandler RETURNS TRUE
     */
    @Override
    protected void setupTanks() {
        tanks = new FluidTank[3];
        tanks[0] = new FluidTank(10 * MetalManager.BLOCK_MB);
        tanks[1] = new FluidTank(10 * MetalManager.BLOCK_MB);
        tanks[2] = new FluidTank(10 * MetalManager.BLOCK_MB);
    }

    /**
     * Which tanks can input
     *
     * MUST OVERRIDE IN CHILD CLASSES IF isFluidHandler RETURNS TRUE
     *
     * @return An array with the indexes of the input tanks
     */
    @Override
    protected int[] getInputTanks() {
        return new int[] { INPUT_TANK_1, INPUT_TANK_2 };
    }

    /**
     * Which tanks can output
     *
     * MUST OVERRIDE IN CHILD CLASSES IF isFluidHandler RETURNS TRUE
     *
     * @return An array with the indexes of the output tanks
     */
    @Override
    protected int[] getOutputTanks() {
        return new int[] { OUTPUT_TANK };
    }

    /**
     * Returns true if the given fluid can be inserted
     *
     * More formally, this should return true if fluid is able to enter
     *
     * @param fluid
     */
    @Override
    protected boolean canFill(Fluid fluid) {
        if(fluid == null)
            return false;
        if(tanks[INPUT_TANK_1].getFluid() == null && tanks[INPUT_TANK_2].getFluid() == null)
            return ((AlloyerRecipeHandler)RecipeManager.getHandler(RecipeManager.RecipeType.ALLOYER)).isValidSingle(new FluidStack(fluid, 1000));
        else {
            if(tanks[INPUT_TANK_1].getFluid() != null && fluid == tanks[INPUT_TANK_1].getFluid().getFluid())
                return true;
            else if(tanks[INPUT_TANK_2].getFluid() != null && fluid == tanks[INPUT_TANK_2].getFluid().getFluid())
                return true;
            else if(tanks[INPUT_TANK_1].getFluid() != null &&
                    RecipeManager.getHandler(RecipeManager.RecipeType.ALLOYER).isValidInput(Pair.of(new FluidStack(fluid, 1000), tanks[INPUT_TANK_1].getFluid())))
                return true;
            else if(tanks[INPUT_TANK_2].getFluid() != null &&
                    RecipeManager.getHandler(RecipeManager.RecipeType.ALLOYER).isValidInput(Pair.of(new FluidStack(fluid, 1000), tanks[INPUT_TANK_2].getFluid())))
                return true;
        }
        return false;
    }

    /*******************************************************************************************************************
     * Misc Methods                                                                                                    *
     *******************************************************************************************************************/

    /**
     * Return the container for this tile
     *
     * @param id       Id, probably not needed but could be used for multiple guis
     * @param player   The player that is opening the gui
     * @param world The world
     * @param x        X Pos
     * @param y        Y Pos
     * @param z        Z Pos
     * @return The container to open
     */
    @Override
    public Object getServerGuiElement(int id, EntityPlayer player, World world, int x, int y, int z) {
        return new ContainerAlloyer(player.inventory, this);
    }

    /**
     * Return the gui for this tile
     *
     * @param id       Id, probably not needed but could be used for multiple guis
     * @param player   The player that is opening the gui
     * @param world The world
     * @param x        X Pos
     * @param y        Y Pos
     * @param z        Z Pos
     * @return The gui to open
     */
    @Override
    public Object getClientGuiElement(int id, EntityPlayer player, World world, int x, int y, int z) {
        return new GuiAlloyer(player, this);
    }

    /**
     * Used to get the description to display on the tab
     *
     * @return The long string with the description
     */
    @Override
    public String getDescription() {
        return "" +
                GuiColor.GREEN + GuiTextFormat.BOLD + GuiTextFormat.UNDERLINE + ClientUtils.translate("neotech.text.stats") + ":\n" +
                GuiColor.YELLOW + GuiTextFormat.BOLD + ClientUtils.translate("neotech.text.energyUsage") + ":\n" +
                GuiColor.WHITE + "  " + ClientUtils.formatNumber(getEnergyCostPerTick()) + " RF/tick\n" +
                GuiColor.YELLOW + GuiTextFormat.BOLD + ClientUtils.translate("neotech.text.processTime") + ":\n" +
                GuiColor.WHITE + "  " + getCookTime() + " ticks\n" +
                GuiColor.YELLOW + GuiTextFormat.BOLD + ClientUtils.translate("neotech.text.operations") + ":\n" +
                GuiColor.WHITE + "  " + getModifierForCategory(IUpgradeItem.ENUM_UPGRADE_CATEGORY.MEMORY) + "\n\n" +
                GuiColor.WHITE + ClientUtils.translate("neotech.alloyer.desc") + "\n\n" +
                GuiColor.GREEN + GuiTextFormat.BOLD + GuiTextFormat.UNDERLINE + ClientUtils.translate("neotech.text.upgrade") + ":\n" + GuiTextFormat.RESET +
                GuiColor.YELLOW + GuiTextFormat.BOLD + ClientUtils.translate("neotech.text.processors") + ":\n" +
                GuiColor.WHITE + ClientUtils.translate("neotech.electricFurnace.processorUpgrade.desc") + "\n\n" +
                GuiColor.YELLOW + GuiTextFormat.BOLD + ClientUtils.translate("neotech.text.memory") + ":\n" +
                GuiColor.WHITE + ClientUtils.translate("neotech.electricFurnace.memoryUpgrade.desc") + "\n\n" +
                GuiColor.YELLOW + GuiTextFormat.BOLD + ClientUtils.translate("neotech.text.psu") + ":\n" +
                GuiColor.WHITE + ClientUtils.translate("neotech.electricFurnace.psuUpgrade.desc") + "\n\n" +
                GuiColor.YELLOW + GuiTextFormat.BOLD + ClientUtils.translate("neotech.text.control") + ":\n" +
                GuiColor.WHITE + ClientUtils.translate("neotech.electricFurnace.controlUpgrade.desc") + "\n\n" +
                GuiColor.YELLOW + GuiTextFormat.BOLD + ClientUtils.translate("neotech.text.network") + ":\n" +
                GuiColor.WHITE +  ClientUtils.translate("neotech.electricFurnace.networkUpgrade.desc");
    }

    /**
     * Used to output the redstone single from this structure
     *
     * Use a range from 0 - 16.
     *
     * 0 Usually means that there is nothing in the tile, so take that for lowest level. Like the generator has no energy while
     * 16 is usually the flip side of that. Output 16 when it is totally full and not less
     *
     * @return int range 0 - 16
     */
    @Override
    public int getRedstoneOutput() {
        return isActive() ? 16 : 0;
    }

    /**
     * Used to get what particles to spawn. This will be called when the tile is active
     */
    @Override
    public void spawnActiveParticles(double x, double y, double z) {
        world.spawnParticle(EnumParticleTypes.SMOKE_NORMAL, x, y, z, 0, 0, 0);
        world.spawnParticle(EnumParticleTypes.SMOKE_NORMAL, x, y, z, 0, 0, 0);
        world.spawnParticle(EnumParticleTypes.SMOKE_NORMAL, x, y, z, 0, 0, 0);
    }
}
